p2p/peer.py

```diff
@@ -799,6 +804,40 @@ def __init__(self,
                 # `event.broadcast_config()` API.
                 self.event_bus.broadcast(PeerCountResponse(len(self)), req.broadcast_config())
 
+    async def maybe_connect_more_peers(self) -> None:
+        discovery_interval = 2
+            
+        while self.is_operational:
+            await self.sleep(discovery_interval)
+
+            left_peer_count = self.max_peers - len(self)
+            if left_peer_count > 0:
+                try:
+                    response = await self.wait(
```

Comment:
I wanted to point something out that may not be obvious.

The previous implementation of this started with `Discovery.maybe_connect_to_more_peers` which calls `DiscoveryProto.get_nodes_to_connect` which depending on the protocol calls `kademlia.RoutingTable.get_random_nodes()` or in the case of v5, uses the topics table to return nodes.

https://github.com/ethereum/py-evm/blob/master/p2p/kademlia.py#L236-L257

https://github.com/ethereum/py-evm/blob/master/p2p/discovery.py#L917-L923

You'll notice that both of these implementations use generators to return the list of candidates, and that these generators are lazily consumed and sometimes discarded before being fully consumed.

If I'm reading this correctly, this PR will change this functionality so that the full iterable of candidates is no longer lazily generated.  I don't think that it's a very big issue but I wanted to make note of it.

Reply:
Thanks for pointing this out.

Previously `connect_to_nodes` was given a generator

https://github.com/ethereum/py-evm/blob/b10f1edc4a9a316f72ce3905a62219ba6ca79b90/p2p/discovery.py#L983-L984

 and consumed it lazily as seen here.

https://github.com/ethereum/py-evm/blob/b10f1edc4a9a316f72ce3905a62219ba6ca79b90/p2p/peer.py#L938-L948

The *lazy* aspect of this basically comes into play if the pool becomes *full* e.g. after connecting to the first node. Then, even if the generator was meant to yield e.g. 5 nodes, it would stop being consumed after the first one.

We can not have this under the new design (unless we want to make lazy lahja based cross process generators a thing but honestly that will end up being inefficient as hell) because we are requesting e.g. 5 nodes from another process so that other process needs to consume all 5 of them, wrap them in a tuple and send them over the to that other process that requested them.

But as you said, I don't think that will be an issue. Getting a bunch of nodes out of the routing table shouldn't be a costly operation *I guess*.